[IMP][10.0] Fix permissions

seb-elico opened the pull request:
Before merging this PR, the image can be tested by pulling:
`docker pull sebelico/odoo-docker:10.0-fix-permissions`

This PR contains:

- [IMP] mapping with the host user (environment variable `TARGET_ID`)
- [REF] move `startup.sh` to sources folder for consistency
- [REF] move `scripts` folder for consistency with the folder structure inside the container
- [FIX] when no addons to download, the `grep` function was failing in the boot script due to the `^` char. We created a Python script to workaround this issue, but actually it works if the `grep` is run in another bash script (see `FIXME` in `no_addons.sh`)
- [IMP] create a function to replace the parameters in odoo.conf (will be useful if we plan to add new environment variable for the parameter substitution)
- [IMP] dynamic setup of SSH key
- [IMP] start the `startup.sh` script before doing any operation with odoo user
- [IMP] `ssh` and `scripts` folders/volumes can belong to target user on the host
- [FIX] It's now possible to specify a branch in `ADDONS_REPO` using the same syntax as `oca_dependencies.txt`
- [IMP] Allow multiple addons repos download from an `oca_dependencies.txt` file at the root of the `additional_addons` folder rather than using env variable `ADDONS_REPO`
- [IMP] Added roadmap in README.md

seb-elico commented:
@Elico-Corp/docker-reviewers please review :+1: 

@victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr feel free to comment!

seb-elico commented:
@elicoidal before we merge this PR, I will pull `sebelico/odoo-docker:10.0-fix-permissions` on Superman and use it for all our v10 trunk instances, so that we have a little more REX about the new version, since there are a lot of changes (I just want to make sure we won't break anything)

@victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr feel free to comment!

elicoidal commented:
Later on we need to sit down to coordinate the changes you made and the 
installation script

seb-elico commented:
@elicoidal I found a bug in the new image: when the volumes are not mapped (I think `extra_addons`, `sessions` and `odoo_data`), they will be created by Docker in its internal folders (e.g. `/var/lib/docker/volumes`) as anonymous volumes. But they will belong to `root:root`, meaning that `odoo` user (`1000:1000`) will not be able to write inside (which is a big issue for the sessions and attachments). Workaround is to map those volumes in `d-c.yml` and make sure the folders belong to `1000:1000`, but I definitely need to fix this before we merge!

elicoidal commented:
I hit this bug during the week endÉric CaudalOn Jun 12, 2017 7:26 PM, Sébastien Maillard - Elico Corp <notifications@github.com> wrote:@elicoidal I found a bug in the new image: when the volumes are not mapped (I think extra_addons, sessions and odoo_data), they will be created by Docker in its internal folders (e.g. /var/lib/docker/volumes) as anonymous volumes. But they will belong to root:root, meaning that odoo user (1000:1000) will not be able to write inside (which is a big issue for the sessions and attachments). Workaround is to map those volumes in d-c.yml and make sure the folders belong to 1000:1000, but I definitely need to fix this before we merge!

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

seb-elico commented:
@elicoidal all ownership bugs are fixed, I also fixed the download of repo when specifying the branch in `d-c.yml` (e.g. `- "ADDONS_REPO=git@github.com:Elico-Corp/yugong-odoo.git 10.0"`). Last but not least, I implemented the download of multiple repositories using a local `oca_dependencies.txt` file. Refer to the PR description for full list of changes.

Still need to pull the image on my laptop and Superman for the final tests before we merge.

@victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr feel free to comment! :+1:

elicoidal commented on README.md:
```diff
@@ -3,4 +3,26 @@ Docker files to build dockers based on:
+  to change any of them dynamically.
+  1. Suggested solution:
+    * pattern for ENV variable: "ODOO_CONF_" + parameter + "=" + value, e.g.:
+      * ODOO_CONF_dbfilter=^stable_.*
```
add blank lines before the lists

seb-elico replied:
It's MD, not RST!

elicoidal commented on README.md:
```diff
@@ -3,4 +3,26 @@ Docker files to build dockers based on:
+This work is original based on the docker from http://www.xcg-consulting.fr and available here:
+https://hub.docker.com/r/xcgd/odoo/
+
+# Roadmap
```
can you add manual for the current usage?

seb-elico replied:
There's a separate PR from Noah about that

elicoidal commented on README.md:
```diff
@@ -3,4 +3,26 @@ Docker files to build dockers based on:
+    * either:
+      1. append a user defined odoo.conf to the one in the image (add this user defined
+      odoo.conf by Dockerfile or by volume)
+      2. or use ENV variables to customize parameters (I'm more in favor of that solution)
```
OK with this solution

seb-elico commented:
@victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr any feedback?

seb-elico commented:
@elicoidal all tests are successful and image has been running on our test servers for several days now. I think it's time to merge this PR and backport it to v7~9. Unless @victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr have some feedback, but they didn't reply for the last 13 days now

elicoidal commented:
Go ahead @seb-elico

victormartinelicocorp commented:
> @elicoidal all tests are successful and image has been running on our test servers for several days now. I think it's time to merge this PR and backport it to v7~9. Unless @victormartinelicocorp @Zhouhao12345 @RawEvan @Wizr have some feedback, but they didn't reply for the last 13 days now

@seb-elico Talking by me. Here on this PR there is many commands that ever seem so for review need to study, so I feel this PR is good enough as you tested. But I could not say I understand everything you fix.

RawEvan commented on Dockerfile:
```diff
@@ -81,18 +93,26 @@ RUN git clone https://github.com/OCA/OCB.git -b 10.0 odoo && \
+# Provide read/write access to group (for host user mapping). For some reason, the
+# files added in a volume (e.g. odoo.conf) belong to root. This command imperatively
+# needs to run before creating the volumes.
+RUN chmod -R 775 /opt/odoo && chown -R odoo:odoo /opt/odoo
```
This command will take a lot of time and maybe it could be merged to line 85.
* `chmod` could be replaced by `mkdir -m`
* `chown` is unnecessary according to comments above line85

seb-elico replied:
@RawEvan thanks for your comment. Actually, on line 85, the **folders** are created with user `odoo`. But adding some **files** in those folders (such as `odoo.conf`, see comment on line 99) will set the ownership to `root` for those files. Hence the `chown`.

As for the `chmod`, it's necessary for the same reason (using `mkdir -m` will not ensure the files will also have `775` mode).

RawEvan commented on auto_addons/no_addons.sh:
```diff
@@ -0,0 +1,15 @@
+grep -P "^addons_path\s*=" $1
+if [ $? -ne 0 ]; then
+    echo "addons_path = /opt/odoo/sources/odoo/addons" >> $1
+fi
```
No new line at the end of file.

seb-elico replied:
Empty line at the end of a file is a PEP8 convention so it only applies for `.py`

RawEvan commented on bin/boot:
```diff
@@ -83,17 +105,23 @@ function on_exit() {
 done
 
-wait
+wait
```
No new line at the end of file.

seb-elico replied:
Empty line at the end of a file is a PEP8 convention so it only applies for `.py`

seb-elico commented:
@RawEvan thanks for your comments, please check my answers!
@elicoidal all comments have been answered, let me know when it's merged so that I can backport to previous versions
